namespaced-openvpn

```diff
@@ -157,40 +159,71 @@ def setup_dns(namespace, dns_type):
+    return any(variables)
 
 def route_up(args):
     """Active ingredient: set up routing and DNS in the new namespace."""
     namespace, dns_type, preexisting_routeup = args
 
     dev = os.getenv('dev')
-    # this is the local IP assigned to the tun adapter
+
+    if not dev:
+        raise ValueError("Missing dev environment variable")
+
+    # this is the local IPv4 assigned to the tun adapter
     ifconfig_local = os.getenv('ifconfig_local')
     # this is the default gateway for the tun adapter (typically private IP space)
     route_vpn_gateway = os.getenv('route_vpn_gateway')
-    if not all((dev, ifconfig_local, route_vpn_gateway)):
-        raise ValueError(
-            "Bad options pushed from server",
-            dev, ifconfig_local, route_vpn_gateway
-        )
-    peer_addr = '%s/32' % (route_vpn_gateway,)
 
-    # transfer the tunnel interface and set it to UP
+    ifconfig_ipv6_local = os.getenv('ifconfig_ipv6_local')
+    ifconfig_ipv6_remote = os.getenv('ifconfig_ipv6_remote')
+    ifconfig_ipv6_netbits = os.getenv('ifconfig_ipv6_netbits')
+
+    have_ipv4 = assert_all_or_none(
+        "Bad ipv4 options pushed from server", ifconfig_local, route_vpn_gateway
+    )
+    have_ipv6 = assert_all_or_none(
+        "Bad ipv6 options pushed from server",
+        ifconfig_ipv6_local, ifconfig_ipv6_remote, ifconfig_ipv6_netbits
+    )
+
+    if not have_ipv4 and not have_ipv6:
+        raise ValueError("Got neither ipv4 nor ipv6 configuration from server")
+
     target = namespace if namespace != '' else '1'
     subprocess.check_call([IP_CMD, 'link', 'set', dev, 'netns', target])
     subprocess.check_call(
         _enter_namespace_cmd(namespace) +
         [IP_CMD, 'link', 'set', dev, 'up']
     )
-    # give it its address
-    subprocess.check_call(
-        _enter_namespace_cmd(namespace) +
-        [IP_CMD, 'addr', 'add', ifconfig_local, 'peer', peer_addr, 'dev', dev]
-    )
-    # route all traffic over the tunnel
-    subprocess.check_call(
-        _enter_namespace_cmd(namespace) +
-        [IP_CMD, 'route', 'add', 'default', 'via', route_vpn_gateway, 'src', ifconfig_local]
-    )
+
+    if have_ipv4:
+        peer_addr = '%s/32' % (route_vpn_gateway,)
+        # transfer the tunnel interface and set it to UP
+        # give it its ipv4 address
+        subprocess.check_call(
+            _enter_namespace_cmd(namespace) +
+            [IP_CMD, 'addr', 'add', ifconfig_local, 'peer', peer_addr, 'dev', dev]
+        )
+        # route all traffic over the tunnel
+        subprocess.check_call(
+            _enter_namespace_cmd(namespace) +
+            [IP_CMD, 'route', 'add', 'default', 'via', route_vpn_gateway, 'src', ifconfig_local]
+        )
+
+    if have_ipv6:
+        subprocess.check_call(
+            _enter_namespace_cmd(namespace) +
+            [IP_CMD, 'addr', 'add', ifconfig_ipv6_local, 'peer',
+            ('%s/%s' % (ifconfig_ipv6_remote, ifconfig_ipv6_netbits)), 'dev', dev]
```

Comment:
Seems like this should be `ipconfig_ipv6_local` instead of `ifconfig_ipv6_remote`?

Reply:
Why? The interface itself gets the local address but the peer address is the address is the remote.